src/main/scala/pbdirect/PBFieldWriter.scala

```diff
@@ -25,140 +25,133 @@ import java.io.ByteArrayOutputStream
 
 import cats.{Contravariant, Functor}
 import com.google.protobuf.CodedOutputStream
-import shapeless.{:+:, ::, CNil, Coproduct, Generic, HList, HNil, Inl, Inr, Lazy}
 import enumeratum.values.IntEnumEntry
 
-trait PBWriter[A] {
+trait PBFieldWriter[A] {
   def writeTo(index: Int, value: A, out: CodedOutputStream): Unit
 }
 
-trait LowPriorityPBWriterImplicits {
-  def instance[A](f: (Int, A, CodedOutputStream) => Unit): PBWriter[A] =
-    new PBWriter[A] {
+trait LowPriorityPBFieldWriterImplicits {
+  def instance[A](f: (Int, A, CodedOutputStream) => Unit): PBFieldWriter[A] =
+    new PBFieldWriter[A] {
       override def writeTo(index: Int, value: A, out: CodedOutputStream): Unit =
         f(index, value, out)
     }
-  implicit val hnilWriter: PBWriter[HNil] = instance { (_: Int, _: HNil, _: CodedOutputStream) =>
-    ()
-  }
-  implicit def consWriter[H, T <: HList](
-      implicit head: PBWriter[H],
-      tail: Lazy[PBWriter[T]]): PBWriter[H :: T] =
-    instance { (index: Int, value: H :: T, out: CodedOutputStream) =>
-      head.writeTo(index, value.head, out)
-      tail.value.writeTo(index + 1, value.tail, out)
-    }
-  implicit def prodWriter[A, R <: HList](
-      implicit gen: Generic.Aux[A, R],
-      writer: Lazy[PBWriter[R]]): PBWriter[A] =
-    instance { (index: Int, value: A, out: CodedOutputStream) =>
-      val buffer = new ByteArrayOutputStream()
-      val pbOut  = CodedOutputStream.newInstance(buffer)
-      writer.value.writeTo(1, gen.to(value), pbOut)
-      pbOut.flush()
-      out.writeByteArray(index, buffer.toByteArray)
-    }
-
-  implicit val cnilWriter: PBWriter[CNil] = instance { (_: Int, _: CNil, _: CodedOutputStream) =>
-    throw new Exception("Can't write CNil")
-  }
-  implicit def cconsWriter[H, T <: Coproduct](
-      implicit head: PBWriter[H],
-      tail: PBWriter[T]): PBWriter[H :+: T] =
-    instance { (index: Int, value: H :+: T, out: CodedOutputStream) =>
-      value match {
-        case Inl(h) => head.writeTo(index, h, out)
-        case Inr(t) => tail.writeTo(index, t, out)
+  implicit def embeddedMessageFieldWriter[A](
+      implicit messageWriter: PBMessageWriter[A]): PBFieldWriter[A] =
+    instance { (index, message, out) =>
+      {
+        val buffer    = new ByteArrayOutputStream()
```

Comment:
Shouldn't you close this resource?

Reply:
Ordinarily, yes, you should close all resources. But ByteArrayOutputStream is a bit of a special case. It doesn't allocate any resources (other than a byte array on the heap, which will be garbage collected), as it's just a wrapper around a byte array.

The [Javadoc](https://docs.oracle.com/javase/8/docs/api/java/io/ByteArrayOutputStream.html#close--) states "Closing a ByteArrayOutputStream has no effect"